### PR TITLE
feat: added publish profile handler

### PR DESCRIPTION
**Pull request by wawrzonnn**

### CODE REVIEW

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task goal
Clicking on "Publish profile" button should make this profile visible on main, profiles list in homepage.

### Requirements
1. Clicking on "Publish profile" button should send an API request to our backend, and set `isPublished` to true.
2. Below endpoint should be used: `http://localhost:3000/api/profiles/:id/publish`

### Code Review

#### app/api/profiles/[id]/publish/route.ts
The added code in this file seems to meet the requirements. It exports an async function `POST` which takes a `request` and `id` as parameters. It uses the `authorizeUser` function to authorize the user and then calls the `updateUserData` function with the `id` and `{ isPublished: true }` as parameters. If successful, it returns a JSON response with a success message and the updated user profile. If an error occurs, it returns a response with a message "Something Went Wrong". 

#### src/components/MyProfile/ProfileMain/ProfileMain.tsx
The changes in this file are only removal of console logs. It does not affect the task requirements.

#### src/components/MyProfile/ProfileTopBar/ProfileTopBar.tsx
The changes in this file meet the requirements. The `handlePublishClick` function is added which sends a POST request to the `/api/profiles/:id/publish` endpoint when the "Publish profile" button is clicked. The function checks if the `userId` exists, if not it logs an error message and returns. If the `userId` exists, it sends a POST request with `{ isPublished: true }` in the body. If the response is ok, it logs a success message, otherwise it logs an error message.

However, there's one issue with this file. The "Edit" button's onClick event handler is set to `() => console.log('click')`. This seems like a placeholder and should be replaced with an actual function.

In conclusion, the code changes cover the task requirements except for the "Edit" button's onClick event handler in `src/components/MyProfile/ProfileTopBar/ProfileTopBar.tsx`.

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task goal
Clicking on "Publish profile" button should make this profile visible on main, profiles list in homepage.

### Requirements
1. Clicking on "Publish profile" button should send an API request to our backend, and set `isPublished` to true.
2. Below endpoint should be used: `http://localhost:3000/api/profiles/:id/publish`

### Code Review

#### app/api/profiles/[id]/publish/route.ts
The added code in this file meets the requirements. It creates a new route that updates the `isPublished` field of a user profile to `true`. The route uses the `updateUserData` function from the profile service, which is expected to handle the database update operation.

#### src/components/MyProfile/ProfileMain/ProfileMain.tsx
The changes in this file are only removals of console.log statements. These changes do not affect the task requirements.

#### src/components/MyProfile/ProfileTopBar/ProfileTopBar.tsx
The changes in this file meet the requirements. The "Publish profile" button now sends a POST request to the `/api/profiles/:id/publish` endpoint when clicked. The user ID is correctly retrieved from the session data.

However, there is a minor issue with the request body. The task requirements do not specify that the request body should contain `{ isPublished: true }`. The server-side code does not use the request body to update the `isPublished` field. It sets `isPublished` to `true` directly. Therefore, sending `{ isPublished: true }` in the request body is unnecessary.

### Conclusion
The code changes cover the task requirements. The "Publish profile" button now sends a POST request to the correct endpoint and makes the profile visible on the main profiles list. However, there is a minor issue with the request body that does not affect the functionality but could be improved for clarity and efficiency.

**Review by NerdbordMentor — APPROVED** [review]
## Code Review Report

### Task goal
Clicking on "Publish profile" button should make this profile visible on main, profiles list in homepage.

### Requirements
1. Clicking on "Publish profile" button should send an API request to our backend, and set `isPublished` to true.
2. Below endpoint should be used:
```
curl -X POST \
  http://localhost:3000/api/profiles/:id/publish
```

### Code Review

#### app/api/profiles/[id]/publish/route.ts
The added code in this file seems to meet the requirements. It defines a POST method for the endpoint `/api/profiles/:id/publish` which updates the `isPublished` field of the user profile to `true`. The `authorizeUser()` function is called before updating the user data, ensuring that only authorized users can publish their profiles.

#### src/components/MyProfile/ProfileMain/ProfileMain.tsx
The changes in this file are minor and seem to be related to cleaning up console logs. They do not directly relate to the task requirements.

#### src/components/MyProfile/ProfileTopBar/ProfileTopBar.tsx
The changes in this file meet the task requirements. The "Publish profile" button now sends a POST request to the `/api/profiles/:id/publish` endpoint when clicked. The user ID is correctly retrieved from the session data. Error handling is also implemented, with console error messages displayed if the user ID is not found or if the request fails.

In conclusion, the provided code changes cover the task requirements. The "Publish profile" button now sends a POST request to the correct endpoint and sets `isPublished` to `true` for the user's profile.

File: src/components/MyProfile/ProfileTopBar/ProfileTopBar.tsx [hunk 1]
```diff
@@ -2,20 +2,51 @@
...
+  const router = useRouter()
+
+  const handlePublishClick = async () => {
+    if (!userId) {
+      console.error('No user ID found')
+      return
+    }
+
+    const response = await fetch(
+      `http://localhost:3000/api/profiles/${userId}/publish`,
```
Comment (Sumick) [inline src/components/MyProfile/ProfileTopBar/ProfileTopBar.tsx:22]: **issue:** This is wrong, as it should be profile ID not userId. If you need you can extend session object to retrieve `profileId` as well

File: app/api/profiles/[id]/publish/route.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { NextRequest, NextResponse } from 'next/server'
+import { updateUserData } from '@/backend/profile/profile.service'
+import { authorizeUser } from '@/lib/auth'
+
+export async function POST(request: NextRequest, id: string) {
+  try {
+    await authorizeUser()
+
+    const updatedUser = await updateUserData(id, { isPublished: true })
```
Comment (Sumick) [inline app/api/profiles/[id]/publish/route.ts:9]: **nitpick:** This function should be named `updateUserProfileById`. Currently it suggests that we operate on `User` domain instead of `Profile`

File: src/components/MyProfile/ProfileTopBar/ProfileTopBar.tsx [hunk 3]
```diff
@@ -2,20 +2,51 @@
...
+
+  const router = useRouter()
+
+  const handlePublishClick = async () => {
+    if (!userId) {
+      console.error('No user ID found')
+      return
+    }
+
+    const response = await fetch(
```
Comment (Sumick) [inline src/components/MyProfile/ProfileTopBar/ProfileTopBar.tsx:21]: **nitpick:** We should create something like `lib/apiClient.ts` where we keep all requests to our API instead of declaring them again and again from scratch

**Review by Sumick — CHANGES_REQUESTED** [review]
We need few improvements here in terms of architecture, and maintainability

File: src/lib/apiClient.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+
+import { Session } from 'next-auth'
+const API_URL = 'http://localhost:3000/api';
+
+export const apiClient = {
+  publishProfile: async (session: Session | null) => {
```
Comment (Sumick) [inline src/lib/apiClient.ts:6]: **nitpick:** I would change it to `publishMyProfile` as it's related to only publishing by authorized user. `publishProfile` would fit better if we pass `profileId` as argument instead of `session`

**Review by Sumick — CHANGES_REQUESTED** [review]
One blocking issue found

**Review by AniaPiWo — APPROVED** [review]

File: src/lib/apiClient.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+const API_URL = 'http://localhost:3000/api';
+
+export const apiClient = {
+  publishMyProfile: async (profileId: string) => {
+    if (!profileId) {
+      throw new Error('No profile ID found');
```
Comment (Sumick) [inline src/lib/apiClient.ts:6]: **nitpick:** You don't have to check it if passed argument is not optional? This `if` is useless here

File: src/components/MyProfile/ProfileTopBar/ProfileTopBar.tsx [hunk 6]
```diff
@@ -2,20 +2,41 @@
...
+import { apiClient } from '@/lib/apiClient'
 
 const ProfileTopBar = () => {
-  const handleButtonClick = () => {
-    console.log('clicked')
+  const router = useRouter()
+  
+  const { data: session } = useSession()
+  const profileId = session?.user?.profileId
+  console.log(profileId);
```
Comment (Sumick) [inline src/components/MyProfile/ProfileTopBar/ProfileTopBar.tsx:15]: **nitpick:** Log to remove

File: src/components/MyProfile/ProfileMain/ProfileMain.tsx [hunk 7]
```diff
@@ -14,14 +14,13 @@ import { AppRoutes } from '@/utils/routes'
...
-
+console.log('sesyjka', session);
   if (!session || !session.user) {
     redirect(AppRoutes.home)
   }
 
   const profile = await getProfileByUserEmail(session.user.email)
-  console.log('profilek', profile)
-  console.log(session)
+  console.log('profilek', profile);
```
Comment (Sumick) [inline src/components/MyProfile/ProfileMain/ProfileMain.tsx:23]: **nitpick:** Log to remove

**Review by Sumick — CHANGES_REQUESTED** [review]
Last few things

File: src/components/MyProfile/ProfileMain/ProfileMain.tsx [hunk 8]
```diff
@@ -14,14 +14,12 @@ import { AppRoutes } from '@/utils/routes'
 
 const ProfileMain = async () => {
   const session = await getServerSession(authOptions)
-
+  console.log('sesyjka', session)
```
Comment (Sumick) [inline src/components/MyProfile/ProfileMain/ProfileMain.tsx:17]: **nitpick:** Log to remove

**Review by Sumick — APPROVED** [review]
Great job! One log to remove before merging
